### ISSUE TITLE
Encoding of long strings

### ISSUE THREAD

**Issue by mmontone** [equill/neo4cl#22]
Hi.

I'm getting "Unhandled string" when passing a long string to the queries.
The problem in the encode-string function, the last case of the cond. It says: (< strlen 65535), but it should say (> strlen 65535).

Also, according to the table in this page: https://neo4j.com/docs/bolt/current/packstream/

![image](https://user-images.githubusercontent.com/436110/221072622-4371e9c4-bb1a-43cc-9085-e00227d43d40.png)

Long strings are encoded with `d3`, but in neo4cl code `d2` is used. I don't know if this is a problem of packstream versions, or not.

I changed the encoding, but still doesn't work (my query hangs). So, I'm not able to provide a PR with a fix, just mention this for a review of the current code.

**Comment by equill** [comment]
Thanks for letting me know!

The documentation does say D3 in two places, and D2 in another. I grappled with this in #17, and confirmed that D2 is what the server uses to encode long strings. Good point, though: I've contacted them and asked them to fix the docs.

That's really strange, though: I'm looking at the code, and that `cond` clause definitely reads `(> 65535 strlen)` here - I copy-pasted that direct from my editor. I've also checked that there aren't any un-pushed changes in my local repo, so it's not a bug I fixed but forgot to upload.

The even stranger part is that I ran into #17 from being able to encode and upload large (32-bit length) strings, but failing to _decode_ them when the server returned them. So I'm even more confused than you are.

If you can change the comparison from `<` to `>`, and switch the marker back to `#xd2`, it _should_ start working for you, while I figure out how that bug even got into your copy, and how to fix that.

What version of the code are you using? Is this from Quicklisp, or the latest from here?

**Comment by mmontone** [comment]
> That's really strange, though: I'm looking at the code, and that `cond` clause definitely reads `(> 65535 strlen)` here - I copy-pasted that direct from my editor. I've also checked that there aren't any un-pushed changes in my local repo, so it's not a bug I fixed but forgot to upload.

Yes, that's what I got. The `>` needs to be changed to `<`. 

>  If you can change the comparison from < to >, and switch the marker back to #xd2, it should start working for you, while I figure out how that bug even got into your copy, and how to fix that.

Unfortunately when I do that, my query hangs.

Setting *loglevel* to `:debug` I get this:

```
[2023-02-24 09:42:16] DEBUG Attempting to run autocommit transaction with query 'CREATE (obj:Award { id: $id, title: $title, description: $description, last_edit_time: $last_edit_time, last_edit_user: $last_edit_user, editlist: $editlist, search_index_cache: $search_index_cache, comments: $comments, isdeleted: $isdeleted, import_source: $import_source, import_extra_data: $import_extra_data, import_technical_data: $import_technical_data, sub_title: $sub_title, frontend_status: $frontend_status, version: $version, changelog: $changelog, export_status2: $export_status2} )'
[2023-02-24 09:42:16] DEBUG Sending Packstream message to the Bolt server.
[2023-02-24 09:42:16] DEBUG Attempting to parse server response to RUN message
[2023-02-24 09:42:16] DEBUG Attempting to parse the response to an autocommit RUN request.
[2023-02-24 09:42:16] DEBUG Attempting to read a Bolt chunk from the connection stream, using protocol version 4.3
```

Not sure how to debug this, my $changelog parameter is a bit long.

Also, could be nice to add this "long string" encoding and query to the tests if there no ones for this already there.

> What version of the code are you using? Is this from Quicklisp, or the latest from here?

I'm using version from here github. The Lisp expression I used in the issue was not correct. It says `(> 65535 strlen)` in the code (which should be changed).

**Comment by mmontone** [comment]
Not sure how easy this would be to reproduce, but I paste my backtrace just in case it is useful.

My query hangs, so I interrupt from Emacs. Seems to hang waiting from a response from neo4j in NEO4CL::READ-BOLT-CHUNK-FROM-STREAM:

```
Interrupt from Emacs
   [Condition of type SIMPLE-ERROR]

Restarts:
 0: [CONTINUE] Continue from break.
 1: [RETRY] Retry SLIME REPL evaluation request.
 2: [*ABORT] Return to SLIME's top level.
 3: [ABORT] abort thread (#<THREAD "repl-thread" RUNNING {100C3C5723}>)

Backtrace:
  0: ("bogus stack frame")
  1: (SB-SYS:WAIT-UNTIL-FD-USABLE 612 :INPUT NIL NIL)
  2: (SB-IMPL::REFILL-INPUT-BUFFER #<SB-SYS:FD-STREAM for "socket 127.0.0.1:60024, peer: 127.0.0.1:7687" {100F30EE43}>)
  3: (SB-IMPL::INPUT-UNSIGNED-8BIT-BYTE #<SB-SYS:FD-STREAM for "socket 127.0.0.1:60024, peer: 127.0.0.1:7687" {100F30EE43}> T NIL)
  4: (READ-BYTE #<SB-SYS:FD-STREAM for "socket 127.0.0.1:60024, peer: 127.0.0.1:7687" {100F30EE43}> T NIL)
  5: ((:METHOD NEO4CL::READ-BOLT-CHUNK-FROM-STREAM (STREAM NEO4CL:BOLT-PROTOCOL)) #<SB-SYS:FD-STREAM for "socket 127.0.0.1:60024, peer: 127.0.0.1:7687" {100F30EE43}> #<unavailable argument>) [fast-method]
  6: ((:METHOD NEO4CL::READ-BOLT-MESSAGE-FROM-STREAM (STREAM NEO4CL:BOLT-PROTOCOL)) #<SB-SYS:FD-STREAM for "socket 127.0.0.1:60024, peer: 127.0.0.1:7687" {100F30EE43}> #<NEO4CL:BOLT-PROTOCOL-4-3 {100F30F13..
  7: ((:METHOD NEO4CL::PARSE-AUTOCOMMIT-TX-RESPONSE (USOCKET:STREAM-USOCKET NEO4CL:BOLT-PROTOCOL-4-3)) #<unavailable argument> #<unavailable argument>) [fast-method]
  8: ((:METHOD NEO4CL:BOLT-TRANSACTION-AUTOCOMMIT (NEO4CL:BOLT-SESSION STRING)) #<NEO4CL:BOLT-SESSION {100F3261D3}> #<unavailable argument> :PARAMETERS #<unavailable argument>) [fast-method]
  9: ((SB-PCL::EMF NEO4CL:BOLT-TRANSACTION-AUTOCOMMIT) #<unused argument> #<unused argument> #<NEO4CL:BOLT-SESSION {100F3261D3}> "CREATE (obj:Award { id: $id, title: $title, description: $description, last..
 10: (QUERY "CREATE (obj:Award { id: $id, title: $title, description: $description, last_edit_time: $last_edit_time, last_edit_user: $last_edit_user, editlist: $editlist, search_index_cache: $search_index_..
```

If the problem persist, I can paste the whole query + parameters later.

**Comment by equill** [comment]
You're right, and that's embarrassing: the comparison is the wrong way around.

OK, now I have a separate branch for this one: `22_Encoding_of_long_strings`. Again, I've committed what looks like the fix, but I need to also add enough testing to make sure it actually _does_ solve the problem.

At least now I know it's easy to test, by feeding it the text of GPLv3 :)

I'll try to reproduce it from here. Are you using Neo4j 4.3.9, or a more modern version? I've been thinking I really should divert some time to support v5, since that's been out for a little while.

**Comment by mmontone** [comment]
That sounds great. Thank you. 

I'm using 4.3.9 because neo4cl didn't work for neo4j v5. I was about to ask you if you have an idea of what are the things that changed in Neo4j v5 regarding the protocol.

**Comment by equill** [comment]
This looks better. Changes so far:

- Bugfix the length comparison for encoding 32-bit strings.
- Rebase this branch on `main` after merging the branch for #23 into it.
- Add basic sanity-tests for encoding 16- and 32-bit strings.
- Add debug-logging to affirm which string-length is being encoded, as well as decoded.
- Add round-trip testing for strings of all length.
    - Hooray for public-domain literature! The 32-bit test involves round-tripping the entire text of the novel Dracula, thanks to Project Gutenberg. Regrettably, I have not yet invested the time into a mash-up of this with Sherlock Holmes and Pooh Bear, but the test is still valid, if less fun than it could be.

**Comment by equill** [comment]
The bad news: the problem is not fully fixed.

The good news: I can replicate the hang at "Attempting to read a Bolt chunk from the connection stream, using protocol version 4.3".

That happens while sending the string _to_ the server, not while receiving it. And that means that we're tripping over #10 here. so at least the cause isn't a mystery.

**Comment by equill** [comment]
Fixed/implemented #10 and now it doesn't hang any more. So that means both causes of this are fixed, too.

**Comment by equill** [comment]
Merged to `main`, and now we have version `0.2.4`.

Feel free to re-open this if it still hangs for you, but I'm confident that it's fixed now.

**Comment by mmontone** [comment]
I've just re-run an import with some big objects and had no errors. Excellent work!! Thank you.

Btw, I'm using this to encode `T`:

```lisp
(defmethod neo4cl::encode-element ((element (eql t)))
	 (neo4cl::encode-boolean t))
```

You may want to add that perhaps.
